ddl/db_test.go

```diff
@@ -4096,21 +4096,57 @@ func (s *testSerialDBSuite) TestModifyColumnBetweenStringTypes(c *C) {
 	tk.MustExec("create table tt (a varchar(10));")
 	tk.MustExec("insert into tt values ('111'),('10000');")
 	tk.MustExec("alter table tt change a a varchar(5);")
-	c2 := getModifyColumn(c, s.s.(sessionctx.Context), "test", "tt", "a", false)
-	c.Assert(c2.FieldType.Flen, Equals, 5)
+	mvc := getModifyColumn(c, s.s.(sessionctx.Context), "test", "tt", "a", false)
+	c.Assert(mvc.FieldType.Flen, Equals, 5)
 	tk.MustQuery("select * from tt").Check(testkit.Rows("111", "10000"))
 	tk.MustGetErrMsg("alter table tt change a a varchar(4);", "[types:1406]Data Too Long, field len 4, data len 5")
 	tk.MustExec("alter table tt change a a varchar(100);")
+	tk.MustQuery("select length(a) from tt").Check(testkit.Rows("3", "5"))
 
+	// char to char
 	tk.MustExec("drop table if exists tt;")
 	tk.MustExec("create table tt (a char(10));")
 	tk.MustExec("insert into tt values ('111'),('10000');")
+	tk.MustExec("alter table tt change a a char(5);")
+	mc := getModifyColumn(c, s.s.(sessionctx.Context), "test", "tt", "a", false)
+	c.Assert(mc.FieldType.Flen, Equals, 5)
 	tk.MustQuery("select * from tt").Check(testkit.Rows("111", "10000"))
 	tk.MustGetErrMsg("alter table tt change a a char(4);", "[types:1406]Data Too Long, field len 4, data len 5")
+	tk.MustExec("alter table tt change a a char(100);")
+	tk.MustQuery("select length(a) from tt").Check(testkit.Rows("3", "5"))
+
+	// binary to binary
+	tk.MustExec("drop table if exists tt;")
+	tk.MustExec("create table tt (a binary(10));")
+	tk.MustExec("insert into tt values ('111'),('10000');")
+	tk.MustGetErrMsg("alter table tt change a a binary(5);", "[types:1406]Data Too Long, field len 5, data len 10")
+	mb := getModifyColumn(c, s.s.(sessionctx.Context), "test", "tt", "a", false)
+	c.Assert(mb.FieldType.Flen, Equals, 10)
+	tk.MustQuery("select * from tt").Check(testkit.Rows("111\x00\x00\x00\x00\x00\x00\x00", "10000\x00\x00\x00\x00\x00"))
+	tk.MustGetErrMsg("alter table tt change a a binary(4);", "[types:1406]Data Too Long, field len 4, data len 10")
```

Comment:
we can add more tests, for example, blob type which is also binary string; char type too.

Reply:
mysql> alter table t modify b char(55) binary;
Query OK, 0 rows affected, 1 warning (0.01 sec)
Records: 0  Duplicates: 0  Warnings: 1

However, in MySQL the length won't change during alter `binary(22)` to `char(33) binary`

Reply:
This behavior of MySQL is reasonable. 

`char(33) binary` defines the collation to "_bin", the ordering/comparison behavior may changed, but `char` alway right-padded with spaces to the specified length according to [here](https://dev.mysql.com/doc/refman/8.0/en/char.html).  So length is not influenced by collation. (BTW, `char(33) binary` become ambiguous since MySQL 8.0 [provide two `_bin` collations to `utf8mb4`](https://dev.mysql.com/doc/refman/8.0/en/string-type-syntax.html))

For the `char` type, the retrieved value would keep the trailing spaces only if ` SQL_MODE contains PAD_CHAR_TO_FULL_LENGTH`, which is disable at default. That's why `Length()` result not changed. (On the contrary, `binary` padding with 0x00 and always keep the padding bytes at retrieve, so the 0x00 can be count at `Length()`)

If we add `PAD_CHAR_TO_FULL_LENGTH` to sql_mode,  execute `alter table t modify b char(33) binary;`, we can get `length = 33`.

---
But when it comes to TiDB, column change to different charset still unsupported now. So I think maybe it's better to add such tests along with the feature development.

Reply:
Every Impressive, thanks LENSHOOD